src/data/reports/twofactor-monitoring/MonitoringTwoFactorD2Repository.ts

```diff
@@ -0,0 +1,203 @@
+import { d2ToolsNamespace } from "../../common/clients/storage/Namespaces";
+import { StorageClient } from "../../common/clients/storage/StorageClient";
+import { CsvData } from "../../common/CsvDataSource";
+import { CsvWriterDataSource } from "../../common/CsvWriterCsvDataSource";
+import { Instance } from "../../common/entities/Instance";
+import { downloadFile } from "../../common/utils/download-file";
+import { Pagination } from "../mal-data-approval/MalDataApprovalDefaultRepository";
+import { MonitoringTwoFactorOptions } from "../../../domain/reports/twofactor-monitoring/entities/MonitoringTwoFactorOptions";
+import { MonitoringTwoFactorRepository } from "../../../domain/reports/twofactor-monitoring/repositories/MonitoringTwoFactorRepository";
+import { MonitoringTwoFactorUser } from "../../../domain/reports/twofactor-monitoring/entities/MonitoringTwoFactorUser";
+import { paginate } from "../../../domain/common/entities/PaginatedObjects";
+import { MonitoringTwoFactorPaginatedObjects } from "../../../domain/reports/twofactor-monitoring/entities/MonitoringTwoFactorPaginatedObjects";
+import { NamedRef } from "../../../domain/common/entities/Ref";
+
+export class MonitoringTwoFactorD2Repository implements MonitoringTwoFactorRepository {
+    private storageClient: StorageClient;
+
+    constructor(private api: D2Api) {
+        const instance = new Instance({ url: this.api.baseUrl });
+        this.storageClient = new DataStoreStorageClient("user", instance);
+    }
+
+    async get(
+        namespace: string,
+        options: MonitoringTwoFactorOptions
+    ): Promise<MonitoringTwoFactorPaginatedObjects<MonitoringTwoFactorUser>> {
+        const { paging, sorting } = options;
+        const groupId = await this.getTwoFactorUserGroup(namespace);
+        const objects = await this.getInvalidUsers(groupId.id);
+
+        const userGroups = _(objects)
+            .flatMap(object => object.userGroups)
+            .uniqBy("id")
+            .value();
+
+        const filteredRows = await this.getFilteredRows(objects, options);
+
+        const { pager, objects: rowsInPage } = paginate(filteredRows, sorting, paging);
+        return {
+            pager: pager,
+            objects: rowsInPage,
+            users: objects,
+            groups: userGroups,
+        };
+    }
+
+    private async getFilteredRows(
+        objects: MonitoringTwoFactorUser[],
+        options: MonitoringTwoFactorOptions
+    ): Promise<MonitoringTwoFactorUser[]> {
+        const { groups, usernameQuery } = options;
+
+        return objects.filter(row => {
+            const isInGroup = groups.length === 0 || row.userGroups.some(group => groups.includes(group.id));
+
+            const isInSearchQuery = _.includes(row.username, usernameQuery);
+
+            return isInGroup && isInSearchQuery;
+        });
+    }
+    private async getTwoFactorUserGroup(namespace: string): Promise<NamedRef> {
+        const { TWO_FACTOR_GROUP_ID: group } = (await this.api
```

Comment:
Due to d2_tools refactor this should be `twoFactorGroup` when that refactor is pushed.